### PR TITLE
Retrieve instances widget all checkbox

### PR DESCRIPTION
**Pull request by antonrohr**
This adds the functionality for the "all instances" checkbox of the
project selection feature.
http://b/202391472

### CODE REVIEW

File: src/SessionSetup/PersistentStorage.cpp [hunk 1]
```diff
@@ -40,4 +43,24 @@ void SaveProjectToPersistentStorage(std::optional<Project> project) {
...
+    return InstanceListScope::kAllReservedInstances;
+  }
+  return InstanceListScope::kOnlyOwnInstances;
+}
+
+void SaveInstancesScopeToPersistentStorage(InstanceListScope scope) {
+  QSettings settings;
+  switch (scope) {
+    case InstanceListScope::kOnlyOwnInstances:
+      settings.remove(kInstancesScopeKey);
```
Comment (beckerhe) [inline src/SessionSetup/PersistentStorage.cpp:58]: I recommend not to use "key does not exist" as a valid setting here. Because that means you can't decide whether the user deliberately chose `kOnlyOwnInstances` the last time when they started Orbit or if it just the default.

Right now that makes no difference but maybe we wanna change that behaviour in the future and then it's valuable to know if the default should apply (the user never started Orbit) or if we should migrate the user's last choice to the equivalent of the new behaviour.

TLDR: I recommend to use two strings for the enum values. That also allows to add more scopes in the future if needed.

Comment (beckerhe) [reply src/SessionSetup/PersistentStorage.cpp:58]: Ahh I just saw that this is not new code and the setting already existed this way? If that's the case and we can't break it feel free to ignore me on this.

File: src/SessionSetup/RetrieveInstancesWidgetTest.cpp [hunk 2]
```diff
@@ -449,4 +461,51 @@ TEST_F(RetrieveInstancesWidgetTestStarted, ProjectChangeFailed) {
...
+  EXPECT_CALL(mock_retrieve_instances_,
+              LoadInstances(std::optional<Project>(std::nullopt),
+                            InstanceListScope(InstanceListScope::kAllReservedInstances)))
+      .WillOnce(Return(Future<ErrorMessageOr<QVector<Instance>>>(ErrorMessage{"error"})));
+
+  QTest::mouseClick(all_check_box_, Qt::MouseButton::LeftButton);
+
+  QMetaObject::invokeMethod(
+      &widget_, []() { QCoreApplication::exit(); }, Qt::QueuedConnection);
+  QCoreApplication::exec();
```
Comment (beckerhe) [inline src/SessionSetup/RetrieveInstancesWidgetTest.cpp:503]: I'm curious. Why does QCoreApplicaton::processEvents does not work in this case?

File: src/SessionSetup/RetrieveInstancesWidget.cpp [hunk 3]
```diff
@@ -89,15 +89,15 @@ void RetrieveInstancesWidget::Start() {
   InitialLoad(LoadLastSelectedProjectFromPersistentStorage());
 }
 
-InstanceListScope RetrieveInstancesWidget::GetInstanceListScope() const {
+InstanceListScope RetrieveInstancesWidget::GetSelectedInstancesScope() const {
```
Comment (beckerhe) [inline src/SessionSetup/RetrieveInstancesWidget.cpp:91]: The name is a bit confusing to me. I assume it's not about selected instances, right? It's about the selected instance-scope, isn't it? For whatever reason I immediately think about selected instances and I'm worried that others will also be confused.

What about GetSelectedInstanceListScope? Longer I know, but clearer I think.

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/SessionSetup/RetrieveInstancesWidgetTest.cpp [hunk 4]
```diff
@@ -449,4 +461,51 @@ TEST_F(RetrieveInstancesWidgetTestStarted, ProjectChangeFailed) {
...
+              LoadInstances(std::optional<Project>(std::nullopt),
+                            InstanceListScope(InstanceListScope::kOnlyOwnInstances)))
+      .WillOnce(Return(Future<ErrorMessageOr<QVector<Instance>>>(kTestInstancesProject1)));
+
+  QTest::mouseClick(all_check_box_, Qt::MouseButton::LeftButton);
+  QCoreApplication::processEvents();
+
+  VerifyLastLoadingReturnedInstanceList(kTestInstancesProject1All);
+  VerifyAndClearSignalsOfSuccessfulLoadingCycle();
+  VerifyAllElementsAreEnabled();
```
Comment (karupayun) [inline src/SessionSetup/RetrieveInstancesWidgetTest.cpp:479]: Here we are checking 3 methods twice. In previous methods we are checking a subset of them. It's not always very clear the difference. What do you think about having a helper Verify function that could take a few parameters and based on that execute the right Verify methods? Could make these tests easy to understand and to modify.

File: src/SessionSetup/PersistentStorage.cpp [hunk 5]
```diff
@@ -40,4 +43,24 @@ void SaveProjectToPersistentStorage(std::optional<Project> project) {
...
 
+[[nodiscard]] InstanceListScope LoadInstancesScopeFromPersistentStorage() {
+  QSettings settings;
+  if (settings.contains(kInstancesScopeKey)) {
+    return InstanceListScope::kAllReservedInstances;
+  }
+  return InstanceListScope::kOnlyOwnInstances;
+}
+
+void SaveInstancesScopeToPersistentStorage(InstanceListScope scope) {
```
Comment (karupayun) [inline src/SessionSetup/PersistentStorage.cpp:54]: I don't have context her but I found strange this implementation.

1- Depending on the switch removing or rather setting a value. Consider modify the code to have similar functionality in both branches.

2- Also we are creating a local QSettings and not returning it at all. But deducing from the last method, seems to be some common pattern from QT, so I'm 100% sure that this last issue is because lack of context.

File: src/SessionSetup/RetrieveInstancesWidgetTest.cpp (hunk 2 above)
Comment (antonrohr) [reply src/SessionSetup/RetrieveInstancesWidgetTest.cpp:503]: Since `LoadInstances` returns an error, the is a QMessageBox displayed to the user. If I understand correctly: Since the QMessageBox is modal, a new event queue is started and that means processEvents becomes blocking.

Comment (beckerhe) [reply src/SessionSetup/RetrieveInstancesWidgetTest.cpp:503]: Got it. Thanks.

File: src/SessionSetup/PersistentStorage.cpp (hunk 1 above)
Comment (antonrohr) [reply src/SessionSetup/PersistentStorage.cpp:58]: > Because that means you can't decide whether the user deliberately chose kOnlyOwnInstances the last time when they started Orbit or if it just the default.

But this is exactly what I want. At this point I really dont care if it was chosen deliberately, or if its the default. 

I looked a while into making this more sustainable in the sense of saving enum values to QSettings. I couldn't find an elegant and simple solution. Q_ENUMS would probably the best way, but its again more work for enums that are not part of a QObject. 

I decided to go this simple and not so elegant way and am aware of the tech debt, aka that adding more scopes in the future is slightly more work.

File: src/SessionSetup/PersistentStorage.cpp (hunk 5 above)
Comment (antonrohr) [reply src/SessionSetup/PersistentStorage.cpp:54]: Regarding 2. Yes, indeed this is a common pattern. The documentation actually does a good job explaining this in case you are interested. https://doc.qt.io/qt-5/qsettings.html

Regarding the other point. 
Yes that is a bit strange here, but I decided to do it anyways, because a truly elegant solution would mean quite a bit of work. Please see my response to the other comment for more info.

File: src/SessionSetup/RetrieveInstancesWidgetTest.cpp (hunk 4 above)
Comment (antonrohr) [reply src/SessionSetup/RetrieveInstancesWidgetTest.cpp:479]: I encapsulated these 3 calls into `VerifySuccessfulLoadAndUIState` and used it where applicable in the code.

File: src/SessionSetup/RetrieveInstancesWidget.cpp (hunk 3 above)
Comment (antonrohr) [reply src/SessionSetup/RetrieveInstancesWidget.cpp]: done
